drivers: serial: stm32: expand DMAT errata test

rbrscoleman opened the pull request:
I manually checked the errata sheets for all STM32Hxx/STM32Uxx parts to confirm the defect's presence or absence. It appears to have been resolved in silicon on newer parts (e.g., STM32U3xx family), so hopefully this test will not need grow further (knock on wood).

The `SOC_STM32H523XX` Kconfig does not actually exist in `main` yet; it will be introduced by https://github.com/zephyrproject-rtos/zephyr/pull/92353.

etienne-lms commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,18 +25,24 @@ config UART_STM32
-	          Errata Sheet 0565 § 2.14.1 and §2.15.1 for stm32h56x/h57x
+	  setting/clearing DMAT bit" as described in the errata sheets:
+
+	  - ES0499 §2.21.2 and §2.22.1 for STM32U575xx/U585xx
+	  - ES0561 §2.11.2 and §2.12.1 for STM32H503xx
+	  - ES0565 §2.16.2 and §2.17.1 for STM32H562xx/H563xx/H573xx
+	  - ES0621 §2.14.2 and §2.15.1 for STM32H523xx/H533xx
```
I would suggest to mention the ES revision number, for consistency of the section references.
By the way, character `§` is not a standard ASCII char. I would suggest to replace it with explicit `section` word.
```suggestion
	  - ES0499 rev. 10, sections 2.21.2 and 2.22.1 for STM32U575xx/U585xx
	  - ES0561 rev. 2, sections 2.11.2 and 2.12.1 for STM32H503xx
	  - ES0565 rev. 7, sections 2.16.2 and 2.17.1 for STM32H562xx/H563xx/H573xx
	  - ES0621 rev. 2, sections 2.14.2 and 2.15.1 for STM32H523xx/H533xx

etienne-lms reviewed (commented):
LGTM with some suggestions.

erwango reviewed (changes requested):
Thanks for this check!

rbrscoleman commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,18 +25,23 @@ config UART_STM32
 
-if SOC_STM32U575XX || SOC_STM32U585XX || \
-		   SOC_STM32H562XX || SOC_STM32H563XX || SOC_STM32H573XX
+if SOC_STM32H503XX || \
```
What about making the choice “depend on” the condition, rather than wrapping it in an `if`? Then there's no uncertainty about how to label the `endif`.

mathieuchopstm replied:
I'm no Kconfig expert so I may be wrong, but I think the choice**s** symbols *also* have to be `if`-gated (i.e., if you used `depends on` then *all choices* would need it too). This would be much uglier so I'd rather not.

But if only one `depends on` on the `choice` itself works (i.e., nothing needed on the choice**s**), then fine by me.

mathieuchopstm replied:
After reading the *very helpful* Kconfig documentation (sarcasm):
> choices:
>
> ```
> "choice"
> <choice options>
> <choice block>
> "endchoice"
> ```
> This defines a choice group and accepts “prompt”, “default”, “depends on”, and “help” attributes as options.
>
> A choice only allows a single config entry to be selected.

I am *guessing* that it works. So I'll give my +1 on the assumption that you tested a `depends on` on a `choice` propagates to the choice**s** too.

rbrscoleman replied:
Yes, sorry, I should've indicated that I did try it. As far as I can see, `depends on` on the `choice` behaves the same as wrapping the whole thing in `if` (choice options are dependent on the choice, so they inherit its dependencies). I think an `if` is only useful for deduplicating the same dependency criteria across multiple otherwise-unrelated configs.

Specifically, I menuconfig'd the `hello_world` sample for two different Nucleos (so that both boards would pass the `UART_STM32` test), one affected, one not (`west build --board nucleo_f401re samples/hello_world/ --target menuconfig` vs. `west build --board nucleo_h533re samples/hello_world/ --target menuconfig`), and searched (`/`) for the symbol `UART_STM32U5_ERRATA_DMAT` to confirm it and its choices were/weren't hidden. I also tried explicitly selecting an errata workaround method (`west build ... -- -DCONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER=y`) for the NUCLEO-F401RE, and confirmed that it was ignored:

> warning: The choice symbol UART_STM32U5_ERRATA_DMAT_LOWPOWER (defined at
> drivers/serial/Kconfig.stm32:47) was selected (set =y), but no symbol ended up as the choice
> selection.

erwango reviewed (changes requested):
I had pending comments not committed, here they are

rbrscoleman committed:
drivers: serial: stm32: expand DMAT errata test.

I manually checked the errata sheets for all STM32Hxx/STM32Uxx parts to
confirm the defect's presence or absence. It appears to have been resolved
in silicon on newer parts (e.g., STM32U3xx family), so hopefully this test
will not need grow further (knock on wood).

Co-authored-by: Nicolas Zuluaga <nicolas.zuluaga@rbr-global.com>
Signed-off-by: Samuel Coleman <samuel.coleman@rbr-global.com>
